README
MessagePack PHP extension
-----------
Binary-based efficient data interchange format.


*Requirements

  MessagePack PHP Extension is only tested on Linux , but it may run on other
  UNIX-like platforms.

  gcc >= 4.1
  php5-dev >= 5.2

   are required to build.


*Installation

    $ phpize
    $ ./configure
    $ make
    $ sudo make install

  To load extension, copy ini/msgpack.ini to included path and restart apache.

  For example, to install on Debian GNU/Linux, 
    $ cp ini/msgpack.ini /etc/php5/apache2/conf.d/
    $ sudo /etc/init.d/apache2 restart

*Test

  phpunit is required to test.
    $ cd test
    $ make test

*Usage

    see example/sample.php for example.

    $  php sample.php



Copyright (C) 2008-2009 Takayuki Sato

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.



ChangeLog
2009-11-29 Takayuki Sato <armyofpigs@gmail.com>
	
	* 0.1.0: first build.


test/TestCase.php
<?php
require_once('PHPUnit/Framework.php');

function check($length, $obj)
{
  $v = msgpack_pack($obj);
  if (strlen($v) != $length) {
    return false;
  }
  if (msgpack_unpack($v) !== $obj) {
    return false;
  }
  return true;
}

function check_array($overhead, $num)
{
  $a = array();
  for ($i = 0; $i < $num; $i++) {
    $a []= NULL;
  }
  /* php array is assoc list */
  return check(($num * 2 + $overhead), $a);
}
    
function check_raw ($overhead, $num)
{
  $str = "";
  for ($i = 0; $i < $num; $i++) {
    $str .= " ";
  }
  return check($num + $overhead, $str);
}

function match($obj, $buf)
{
  if (msgpack_pack($obj) != $buf) {
    return false;
  }
  if (msgpack_unpack($buf) != $obj) {
    return false;
  }
  return true;
}

class TestCase extends PHPUnit_Framework_TestCase
{
  function test_1()
  {
    $a = array(NULL, True, False, 0, 1,
	       (1 << 6), (1 << 7) - 1, -1, -((1<<5)-1), -(1<<5));
    foreach ($a as $o) {
      $this->assertTrue(check(1, $o));
    }
  }

  function test_2()
  {
    $a = array(1 << 7, (1 << 8) - 1,
	       -((1<<5)+1), -(1<<7));
    foreach ($a as $o) {
      $this->assertTrue(check(2, $o));
    }
  }

  function test_3()
  {
    $a = array(1 << 8, (1 << 16) - 1,
	       -((1<<7)+1), -(1<<15));
    foreach ($a as $o) {
      $this->assertTrue(check(3, $o));
    }
  }

  function test_5()
  {
    $a = array(1 << 16, (1 << 32) - 1,
	       -((1<<15)+1), -(1<<31));
    foreach ($a as $o) {
      $this->assertTrue(check(5, $o));
    }
  }

  function test_9()
  {
    $a = array(1 << 32, -((1<<31)+1),
	       1.0, 0.1, -0.1, -1.0);
    foreach ($a as $k => $o) {
      $this->assertTrue(check(9, $o), $msg);
    }
  }

  function test_fixraw()
  {
    $this->assertTrue(check_raw(1, 0));
    $this->assertTrue(check_raw(1, (1<<5) - 1));
  }

  function test_raw16()
  {
    $this->assertTrue(check_raw(3, (1 << 5)));
    $this->assertTrue(check_raw(3, (1<<16) - 1));
  }

  function test_raw32()
  {
    $this->assertTrue(check_raw(5, 1<<16));
  }

  function test_fixarray()
  {
    $this->assertTrue(check_array(1, 0));
    $this->assertTrue(check_array(1, (1 << 4) - 1));
  }

  function test_array16()
  {
    $this->assertTrue(check_array(3, 1 << 4));
    /* max array elements is 1310077 */
    //    $this->assertTrue(check_array(3, (1<<16)-1));
  }

  function test_array32()
  {
    //    $this->assertTrue(check_array(5, (1<<16)));
  }

  function test_match()
  {
    $cases = array(
		   array(NULL, pack("c", 0xc0)),
		   array(False, pack("c", 0xc2)),
		   array(True, pack("c", 0xc3)),
		   array(0, pack("c", 0x00)),
		   array(127, pack("c", 0x7f)),
		   array(128, pack("c*", 0xcc, 0x80)),
		   array(256, pack("c*", 0xcd, 0x01, 0x00)),
		   array(-1, pack("c", 0xff)),
		   array(-33, pack("c*",  0xd0, 0xdf)),
		   array(-129, pack("c*",  0xd1, 0xff, 0x7f)),
		   array(1.0, pack("c*", 0xcb, 0x3f, 0xf0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)),
		   array(array(), pack("c", 0x80)),
		   array(range(0, 14), pack("c*", 0x8f, 0x00, 0x00, 0x01, 0x01,
					    0x02, 0x02, 0x03, 0x03, 0x04, 0x04,
					    0x05, 0x05, 0x06, 0x06, 0x07, 0x07,
					    0x08, 0x08, 0x09, 0x09, 0x0a, 0x0a,
					    0x0b, 0x0b, 0x0c, 0x0c, 0x0d, 0x0d,
					    0x0e, 0x0e)),
		   array(range(0, 15), pack("c*", 0xde, 0x00, 0x10, 0x00, 0x00,
					    0x01, 0x01, 0x02, 0x02, 0x03, 0x03,
					    0x04, 0x04, 0x05, 0x05, 0x06, 0x06,
					    0x07, 0x07, 0x08, 0x08, 0x09, 0x09,
					    0x0a, 0x0a, 0x0b, 0x0b, 0x0c, 0x0c,
					    0x0d, 0x0d, 0x0e, 0x0e, 0x0f, 0x0f))
		   );
    foreach ($cases as $k => $x) {
      $this->assertTrue(match($x[0], $x[1]), "$k");
    }
  }
}


test/TestPack.php
<?php
require_once('PHPUnit/Framework.php');

function check($data)
{
  if ($data === msgpack_unpack(msgpack_pack($data)))
    return true;
  else
    return false;
}

class TestPack extends PHPUnit_Framework_TestCase
{
  function test_pack()
  {
    $test_data = array(0, 1, 127, 128, 255, 256, 65535, 65536,
		       -1, -32, -33, -128, -129, -32768, -32769,
		       1.0,
		       "", "a", "a"*31, "a"*32,
		       None, True, False,
		       array(), array(array()), array(array(), NULL),
		       array(10 => 0),
		       (1<<23)
		       );
    foreach ($test_data as $td) {
      $this->assertTrue(check($td));
    }
  }
}


test/TestFormat.php
<?php
require_once('PHPUnit/Framework.php');

function check($src, $should)
{
  $index = 0;
  while ($src->remain()) {
    if ($src->unpacker(NULL) !== $should[$index++])
      return false;
  }
  return true;
}

class TestFormat extends PHPUnit_Framework_TestCase
{
  function testSimpleValue()
  {
    $obj = new Msgpack;
    
    $obj->feed(pack("c*", 0xc0, 0xc2, 0xc3));
    $should = array(null, false, true);
    $this->assertTrue(check($obj, $should));
  }

  function testFixnum()
  {
    $obj = new Msgpack;
    
    $obj->feed(pack("c*", 0x00, 0x40, 0x7f, 0xe0, 0xf0, 0xff));
    $should = array(0, 64, 127, -32, -16, -1);
    $this->assertTrue(check($obj, $should));
  }

  function testFixArray()
  {
    $obj = new Msgpack;

    $obj->feed(pack("c*", 0x80, 0x81, 0x00, 0xc0));
    $should = array(array(), array(null));
    $this->assertTrue(check($obj, $should));
  }

  function testFixRaw()
  {
    $obj = new Msgpack;

    $obj->feed(pack("c*", 0xa0, 0xa1, 0x61, 0xa2, 0x62, 0x63, 0xa3, 0x64, 0x65, 0x66));
    $should = array("", "a", "bc", "def");
    $this->assertTrue(check($obj, $should));
  }
  
  function testFixMap()
  {
    $obj = new Msgpack;

    $obj->feed(pack("c*", 0x82,
		    0x00, 0x81, 0x00, 0x00,
		    0x01, 0x81, 0x00, 0x80));
    $should = array(array(
			  0 => array(0),
			  1 => array(
				     0 => array())
			  ));
    $this->assertTrue(check($obj, $should));
  }
  
  function testUnsignedInt()
  {
    $obj = new Msgpack;

    $obj->feed(pack("c*", 0xcc, 0x00, 0xcc, 0x80, 0xcc, 0xff,
		    0xcd, 0x00, 0x00, 0xcd, 0x80, 0x00, 0xcd, 0xff, 0xff,
		    0xce, 0x00, 0x00, 0x00, 0x00, 0xce, 0x80, 0x00, 0x00, 0x00,
		    0xce, 0xff, 0xff, 0xff, 0xff));
    $should = array(0, 128, 255, 0, 32768, 65535, 0, 2147483648, 4294967295);
    $this->assertTrue(check($obj, $should));
  }

  function testSignedInt()
  {
    $obj = new Msgpack;

    $obj->feed(pack("c*", 0xd0, 0x00, 0xd0, 0x80, 0xd0, 0xff, 0xd1, 0x00, 0x00, 0xd1, 0x80, 0x00,
		    0xd1, 0xff, 0xff, 0xd2, 0x00, 0x00, 0x00, 0x00,
		    0xd2, 0x80, 0x00, 0x00, 0x00, 0xd2, 0xff, 0xff, 0xff, 0xff));
    $should = array(0, -128, -1, 0, -32768, -1, 0, -2147483648, -1);
    $this->assertTrue(check($obj, $should));
  }

  function testRaw()
  {
    $obj = new Msgpack;

    $obj->feed(pack("c*", 0xda, 0x00, 0x00, 0xda, 0x00, 0x01, 0x61, 0xda, 0x00, 0x02, 0x61, 0x62,
		    0xdb, 0x00, 0x00, 0x00, 0x00,
		    0xdb, 0x00, 0x00, 0x00, 0x01, 0x61,
		    0xdb, 0x00, 0x00, 0x00, 0x02, 0x61, 0x62));
    $should = array("", "a", "ab", "", "a", "ab");
    $this->assertTrue(check($obj, $should));
  }

  function testMap()
  {
    $obj = new Msgpack;

    $obj->feed(pack("c*", 0xde, 0x00, 0x00,
		    0xde, 0x00, 0x01, 0x00, 0xc2,
		    0xde, 0x00, 0x02, 0x01, 0xc2, 0x00, 0xc2,
		    0xdf, 0x00, 0x00, 0x00, 0x00,
		    0xdf, 0x00, 0x00, 0x00, 0x01, 0x00, 0xc2,
		    0xdf, 0x00, 0x00, 0x00, 0x02, 0x01, 0xc2, 0x00, 0xc2));
    $should = array(
		    array(),
		    array(0 => false),
		    array(1 => false, 0 => false),
		    array(),
		    array(0 => false),
		    array(1 => false, 0 => false));
    $this->assertTrue(check($obj, $should));
  }
}


pack.h
/* MessagePack for PHP packing routine
 *
 * Copyright (C) 2009 Takayuki SATO
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

#include <stddef.h>
#include <stdlib.h>
#include "pack_define.h"

#ifdef __cplusplus
extern "C" {
#endif

typedef struct msgpack_packer {
    char *buf;
    size_t length;
    size_t buf_size;
} msgpack_packer;

typedef struct Packer Packer;

static inline int msgpack_pack_short(msgpack_packer* pk, short d);
static inline int msgpack_pack_int(msgpack_packer* pk, int d);
static inline int msgpack_pack_long(msgpack_packer* pk, long d);
static inline int msgpack_pack_long_long(msgpack_packer* pk, long long d);
static inline int msgpack_pack_unsigned_short(msgpack_packer* pk, unsigned short d);
static inline int msgpack_pack_unsigned_int(msgpack_packer* pk, unsigned int d);
static inline int msgpack_pack_unsigned_long(msgpack_packer* pk, unsigned long d);
static inline int msgpack_pack_unsigned_long_long(msgpack_packer* pk, unsigned long long d);

static inline int msgpack_pack_uint8(msgpack_packer* pk, uint8_t d);
static inline int msgpack_pack_uint16(msgpack_packer* pk, uint16_t d);
static inline int msgpack_pack_uint32(msgpack_packer* pk, uint32_t d);
static inline int msgpack_pack_uint64(msgpack_packer* pk, uint64_t d);
static inline int msgpack_pack_int8(msgpack_packer* pk, int8_t d);
static inline int msgpack_pack_int16(msgpack_packer* pk, int16_t d);
static inline int msgpack_pack_int32(msgpack_packer* pk, int32_t d);
static inline int msgpack_pack_int64(msgpack_packer* pk, int64_t d);

static inline int msgpack_pack_float(msgpack_packer* pk, float d);
static inline int msgpack_pack_double(msgpack_packer* pk, double d);

static inline int msgpack_pack_nil(msgpack_packer* pk);
static inline int msgpack_pack_true(msgpack_packer* pk);
static inline int msgpack_pack_false(msgpack_packer* pk);

static inline int msgpack_pack_array(msgpack_packer* pk, unsigned int n);

static inline int msgpack_pack_map(msgpack_packer* pk, unsigned int n);

static inline int msgpack_pack_raw(msgpack_packer* pk, size_t l);
static inline int msgpack_pack_raw_body(msgpack_packer* pk, const void* b, size_t l);

static inline int msgpack_pack_write(msgpack_packer* pk, const char *data, size_t l)
{
    char* buf = pk->buf;
    size_t bs = pk->buf_size;
    size_t len = pk->length;

    if (len + l > bs) {
        bs = (len + l) * 2;
        buf = erealloc(buf, bs);
        if (!buf) return -1;
    }
    memcpy(buf + len, data, l);
    len += l;

    pk->buf = buf;
    pk->buf_size = bs;
    pk->length = len;
    return 0;
}

#define msgpack_pack_inline_func(name) \
	static inline int msgpack_pack ## name

#define msgpack_pack_inline_func_cint(name) \
	static inline int msgpack_pack ## name

#define msgpack_pack_user msgpack_packer*

#define msgpack_pack_append_buffer(user, buf, len) \
        return msgpack_pack_write(user, (const char*)buf, len)

#include "pack_template.h"

#ifdef __cplusplus
}
#endif


unpack.h
/*
 * MessagePack for PHP packing routine
 *
 * Copyright (C) 2009 Takayuki SATO
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

#define MSGPACK_MAX_STACK_SIZE  (1024)
#include "unpack_define.h"

typedef struct unpack_user {
  int enable_array;
} unpack_user;


#define msgpack_unpack_struct(name) \
	struct template ## name

#define msgpack_unpack_func(ret, name) \
	static inline ret template ## name

#define msgpack_unpack_callback(name) \
	template_callback ## name

#define msgpack_unpack_object zval*

#define msgpack_unpack_user unpack_user


struct template_context;
typedef struct template_context template_context;

static inline msgpack_unpack_object template_callback_root(unpack_user* u)
{
    return NULL;
}

static inline int template_callback_int64(unpack_user* u, int64_t d, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  ZVAL_LONG(*o, d);
  return 0;
}

static inline int template_callback_uint16(unpack_user* u, uint16_t d, msgpack_unpack_object* o)
{
  return template_callback_int64(u, d, o);
}
static inline int template_callback_uint8(unpack_user* u, uint8_t d, msgpack_unpack_object* o)
{
  return template_callback_int64(u, d, o);
}

static inline int template_callback_uint32(unpack_user* u, uint32_t d, msgpack_unpack_object* o)
{
    return template_callback_int64(u, d, o);
}

static inline int template_callback_uint64(unpack_user* u, uint64_t d, msgpack_unpack_object* o)
{
  return template_callback_int64(u, d, o);
}

static inline int template_callback_int32(unpack_user* u, int32_t d, msgpack_unpack_object* o)
{
  return template_callback_int64(u, d, o);
}

static inline int template_callback_int16(unpack_user* u, int16_t d, msgpack_unpack_object* o)
{
  return template_callback_int64(u, d, o);
}

static inline int template_callback_int8(unpack_user* u, int8_t d, msgpack_unpack_object* o)
{
  return template_callback_int64(u, d, o);
}

static inline int template_callback_double(unpack_user* u, double d, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  ZVAL_DOUBLE(*o, d);
  return 0;
}

static inline int template_callback_float(unpack_user* u, float d, msgpack_unpack_object* o)
{
  return template_callback_double(u, d, o);
}

static inline int template_callback_nil(unpack_user* u, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  ZVAL_NULL(*o);
  return 0;
}

static inline int template_callback_true(unpack_user* u, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  ZVAL_BOOL(*o, 1);
  return 0;
}

static inline int template_callback_false(unpack_user* u, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  ZVAL_BOOL(*o, 0);
  return 0;
}

/* msgpack_pack don't serialize array to array. */
static inline int template_callback_array(unpack_user* u, unsigned int n, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  if (u->enable_array) {
    array_init(*o);
  } else {
    ZVAL_NULL(*o);
  }
  return 0;
}

static inline int template_callback_array_item(unpack_user* u, unsigned int current, msgpack_unpack_object* c, msgpack_unpack_object o)
{
  if (u->enable_array) {
    add_index_zval(*c, current, o);
  } else {
    ;
  }
  return 0;
}

static inline int template_callback_map(unpack_user* u, unsigned int n, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  array_init(*o);
  return 0;
}

static inline int template_callback_map_item(unpack_user* u, msgpack_unpack_object* c, msgpack_unpack_object k, msgpack_unpack_object v)
{
  if (Z_TYPE_P(k) == IS_STRING) {
    add_assoc_zval_ex(*c, Z_STRVAL_P(k), Z_STRLEN_P(k), v);
  } else {
    add_index_zval(*c, Z_LVAL_P(k), v);
  }
  return 0;
}

static inline int template_callback_raw(unpack_user* u, const char* b, const char* p, unsigned int l, msgpack_unpack_object* o)
{
  MAKE_STD_ZVAL(*o);
  ZVAL_STRINGL(*o, p, l, 1);
  return 0;
}

#include "unpack_template.h"


php_msgpack.h
#include "pack.h"
#include "unpack.h"
#define MSGPACK_PACK_INITBUFSIZE 1024

#ifndef PHP_MSGPACK_H
#define PHP_MSGPACK_H

extern zend_module_entry msgpack_module_entry;
#define phpext_msgpack_ptr &msgpack_module_entry

#ifdef PHP_WIN32
#define PHP_MSGPACK_API __declspec(dllexport)
#else
#define PHP_MSGPACK_API
#endif

#ifdef ZTS
#include "TSRM.h"
#endif

PHP_MINIT_FUNCTION(msgpack);
PHP_MSHUTDOWN_FUNCTION(msgpack);
PHP_RINIT_FUNCTION(msgpack);
PHP_RSHUTDOWN_FUNCTION(msgpack);
PHP_MINFO_FUNCTION(msgpack);

static zend_class_entry *msgpack_ce;
PHP_FUNCTION(msgpack_pack);
PHP_FUNCTION(msgpack_unpack);

PHP_METHOD(msgpack, packer);
PHP_METHOD(msgpack, unpacker);
PHP_METHOD(msgpack, feed);
PHP_METHOD(msgpack, remain);
PHP_METHOD(msgpack, dump);
PHP_METHOD(msgpack, init);
PHP_METHOD(msgpack, setopt);

#define MSGPACK_MAXBUFSIZE 1024*1024
/* 
  	Declare any global variables you may need between the BEGIN
	and END macros here:     
*/
ZEND_BEGIN_MODULE_GLOBALS(msgpack)
	char unpacker_buf[MSGPACK_MAXBUFSIZE];
ZEND_END_MODULE_GLOBALS(msgpack)


/* In every utility function you add that needs to use variables 
   in php_msgpack_globals, call TSRMLS_FETCH(); after declaring other 
   variables used by that function, or better yet, pass in TSRMLS_CC
   after the last function argument and declare your utility function
   with TSRMLS_DC after the last declared argument.  Always refer to
   the globals in your function as MSGPACK_G(variable).  You are 
   encouraged to rename these macros something shorter, see
   examples in any other php module directory.
*/

#ifdef ZTS
#define MSGPACK_G(v) TSRMG(msgpack_globals_id, zend_msgpack_globals *, v)
#else
#define MSGPACK_G(v) (msgpack_globals.v)
#endif

#endif	/* PHP_MSGPACK_H */



unpack_template.h
/*
 * MessagePack unpacking routine template
 *
 * Copyright (C) 2008-2009 FURUHASHI Sadayuki
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

#ifndef msgpack_unpack_func
#error msgpack_unpack_func template is not defined
#endif

#ifndef msgpack_unpack_callback
#error msgpack_unpack_callback template is not defined
#endif

#ifndef msgpack_unpack_struct
#error msgpack_unpack_struct template is not defined
#endif

#ifndef msgpack_unpack_struct_decl
#define msgpack_unpack_struct_decl(name) msgpack_unpack_struct(name)
#endif

#ifndef msgpack_unpack_object
#error msgpack_unpack_object type is not defined
#endif

#ifndef msgpack_unpack_user
#error msgpack_unpack_user type is not defined
#endif


msgpack_unpack_struct_decl(_stack) {
	msgpack_unpack_object obj;
	size_t count;
	unsigned int ct;

    union {
        size_t curr;
        msgpack_unpack_object map_key;
    };
};

msgpack_unpack_struct_decl(_context) {
	msgpack_unpack_user user;
	unsigned int cs;
	unsigned int trail;
	unsigned int top;
	msgpack_unpack_struct(_stack) stack[MSGPACK_MAX_STACK_SIZE];
};


msgpack_unpack_func(void, _init)(msgpack_unpack_struct(_context)* ctx)
{
	ctx->cs = CS_HEADER;
	ctx->trail = 0;
	ctx->top = 0;
	ctx->stack[0].obj = msgpack_unpack_callback(_root)(&ctx->user);
}

msgpack_unpack_func(msgpack_unpack_object, _data)(msgpack_unpack_struct(_context)* ctx)
{
	return (ctx)->stack[0].obj;
}


msgpack_unpack_func(int, _execute)(msgpack_unpack_struct(_context)* ctx, const char* data, size_t len, size_t* off)
{
	assert(len >= *off);

	const unsigned char* p = (unsigned char*)data + *off;
	const unsigned char* const pe = (unsigned char*)data + len;
	const void* n = NULL;

	unsigned int trail = ctx->trail;
	unsigned int cs = ctx->cs;
	unsigned int top = ctx->top;
	msgpack_unpack_struct(_stack)* stack = ctx->stack;
	msgpack_unpack_user* user = &ctx->user;

	msgpack_unpack_object obj;
	msgpack_unpack_struct(_stack)* c = NULL;

	int ret;

#define push_simple_value(func) \
	if(msgpack_unpack_callback(func)(user, &obj) < 0) { goto _failed; } \
	goto _push
#define push_fixed_value(func, arg) \
	if(msgpack_unpack_callback(func)(user, arg, &obj) < 0) { goto _failed; } \
	goto _push
#define push_variable_value(func, base, pos, len) \
	if(msgpack_unpack_callback(func)(user, \
		(const char*)base, (const char*)pos, len, &obj) < 0) { goto _failed; } \
	goto _push

#define again_fixed_trail(_cs, trail_len) \
	trail = trail_len; \
	cs = _cs; \
	goto _fixed_trail_again
#define again_fixed_trail_if_zero(_cs, trail_len, ifzero) \
	trail = trail_len; \
	if(trail == 0) { goto ifzero; } \
	cs = _cs; \
	goto _fixed_trail_again

#define start_container(func, count_, ct_) \
	if(msgpack_unpack_callback(func)(user, count_, &stack[top].obj) < 0) { goto _failed; } \
	if((count_) == 0) { obj = stack[top].obj; goto _push; } \
	if(top >= MSGPACK_MAX_STACK_SIZE) { goto _failed; } \
	stack[top].ct = ct_; \
	stack[top].curr = 0;	   \
	stack[top].count = count_; \
	/*printf("container %d count %d stack %d\n",stack[top].obj,count_,top);*/ \
	/*printf("stack push %d\n", top);*/ \
	++top; \
	goto _header_again

#define NEXT_CS(p) \
	((unsigned int)*p & 0x1f)

#define PTR_CAST_8(ptr)   (*(uint8_t*)ptr)
#define PTR_CAST_16(ptr)  msgpack_betoh16(*(uint16_t*)ptr)
#define PTR_CAST_32(ptr)  msgpack_betoh32(*(uint32_t*)ptr)
#define PTR_CAST_64(ptr)  msgpack_betoh64(*(uint64_t*)ptr)

	if(p == pe) { goto _out; }
	do {
		switch(cs) {
		case CS_HEADER:
			switch(*p) {
			case 0x00 ... 0x7f:  // Positive Fixnum
				push_fixed_value(_uint8, *(uint8_t*)p);
			case 0xe0 ... 0xff:  // Negative Fixnum
				push_fixed_value(_int8, *(int8_t*)p);
			case 0xc0 ... 0xdf:  // Variable
				switch(*p) {
				case 0xc0:  // nil
					push_simple_value(_nil);
				//case 0xc1:  // string
				//	again_terminal_trail(NEXT_CS(p), p+1);
				case 0xc2:  // false
					push_simple_value(_false);
				case 0xc3:  // true
					push_simple_value(_true);
				//case 0xc4:
				//case 0xc5:
				//case 0xc6:
				//case 0xc7:
				//case 0xc8:
				//case 0xc9:
				case 0xca:  // float
				case 0xcb:  // double
				case 0xcc:  // unsigned int  8
				case 0xcd:  // unsigned int 16
				case 0xce:  // unsigned int 32
				case 0xcf:  // unsigned int 64
				case 0xd0:  // signed int  8
				case 0xd1:  // signed int 16
				case 0xd2:  // signed int 32
				case 0xd3:  // signed int 64
					again_fixed_trail(NEXT_CS(p), 1 << (((unsigned int)*p) & 0x03));
				//case 0xd4:
				//case 0xd5:
				//case 0xd6:  // big integer 16
				//case 0xd7:  // big integer 32
				//case 0xd8:  // big float 16
				//case 0xd9:  // big float 32
				case 0xda:  // raw 16
				case 0xdb:  // raw 32
				case 0xdc:  // array 16
				case 0xdd:  // array 32
				case 0xde:  // map 16
				case 0xdf:  // map 32
					again_fixed_trail(NEXT_CS(p), 2 << (((unsigned int)*p) & 0x01));
				default:
					goto _failed;
				}
			case 0xa0 ... 0xbf:  // FixRaw
				again_fixed_trail_if_zero(ACS_RAW_VALUE, ((unsigned int)*p & 0x1f), _raw_zero);
			case 0x90 ... 0x9f:  // FixArray
				start_container(_array, ((unsigned int)*p) & 0x0f, CT_ARRAY_ITEM);
			case 0x80 ... 0x8f:  // FixMap
				start_container(_map, ((unsigned int)*p) & 0x0f, CT_MAP_KEY);

			default:
				goto _failed;
			}
			// end CS_HEADER


		_fixed_trail_again:
			++p;

		default:
			if((size_t)(pe - p) < trail) { goto _out; }
			n = p;  p += trail - 1;
			switch(cs) {
			//case CS_
			//case CS_
			case CS_FLOAT: {
					union { uint32_t num; char buf[4]; } f;
					f.num = PTR_CAST_32(n);  // FIXME
					push_fixed_value(_float, *((float*)f.buf)); }
			case CS_DOUBLE: {
					union { uint64_t num; char buf[8]; } f;
					f.num = PTR_CAST_64(n);  // FIXME
					push_fixed_value(_double, *((double*)f.buf)); }
			case CS_UINT_8:
				push_fixed_value(_uint8, (uint8_t)PTR_CAST_8(n));
			case CS_UINT_16:
				push_fixed_value(_uint16, (uint16_t)PTR_CAST_16(n));
			case CS_UINT_32:
				push_fixed_value(_uint32, (uint32_t)PTR_CAST_32(n));
			case CS_UINT_64:
				push_fixed_value(_uint64, (uint64_t)PTR_CAST_64(n));

			case CS_INT_8:
				push_fixed_value(_int8, (int8_t)PTR_CAST_8(n));
			case CS_INT_16:
				push_fixed_value(_int16, (int16_t)PTR_CAST_16(n));
			case CS_INT_32:
				push_fixed_value(_int32, (int32_t)PTR_CAST_32(n));
			case CS_INT_64:
				push_fixed_value(_int64, (int64_t)PTR_CAST_64(n));

			//case CS_
			//case CS_
			//case CS_BIG_INT_16:
			//	again_fixed_trail_if_zero(ACS_BIG_INT_VALUE, (uint16_t)PTR_CAST_16(n), _big_int_zero);
			//case CS_BIG_INT_32:
			//	again_fixed_trail_if_zero(ACS_BIG_INT_VALUE, (uint32_t)PTR_CAST_32(n), _big_int_zero);
			//case ACS_BIG_INT_VALUE:
			//_big_int_zero:
			//	// FIXME
			//	push_variable_value(_big_int, data, n, trail);

			//case CS_BIG_FLOAT_16:
			//	again_fixed_trail_if_zero(ACS_BIG_FLOAT_VALUE, (uint16_t)PTR_CAST_16(n), _big_float_zero);
			//case CS_BIG_FLOAT_32:
			//	again_fixed_trail_if_zero(ACS_BIG_FLOAT_VALUE, (uint32_t)PTR_CAST_32(n), _big_float_zero);
			//case ACS_BIG_FLOAT_VALUE:
			//_big_float_zero:
			//	// FIXME
			//	push_variable_value(_big_float, data, n, trail);

			case CS_RAW_16:
				again_fixed_trail_if_zero(ACS_RAW_VALUE, (uint16_t)PTR_CAST_16(n), _raw_zero);
			case CS_RAW_32:
				again_fixed_trail_if_zero(ACS_RAW_VALUE, (uint32_t)PTR_CAST_32(n), _raw_zero);
			case ACS_RAW_VALUE:
			_raw_zero:
				push_variable_value(_raw, data, n, trail);

			case CS_ARRAY_16:
				start_container(_array, (uint16_t)PTR_CAST_16(n), CT_ARRAY_ITEM);
			case CS_ARRAY_32:
				/* FIXME security guard */
				start_container(_array, (uint32_t)PTR_CAST_32(n), CT_ARRAY_ITEM);

			case CS_MAP_16:
				start_container(_map, (uint16_t)PTR_CAST_16(n), CT_MAP_KEY);
			case CS_MAP_32:
				/* FIXME security guard */
				start_container(_map, (uint32_t)PTR_CAST_32(n), CT_MAP_KEY);

			default:
				goto _failed;
			}
		}

_push:
	if(top == 0) { goto _finish; }
	c = &stack[top-1];
	switch(c->ct) {
	case CT_ARRAY_ITEM:
		if(msgpack_unpack_callback(_array_item)(user, c->curr, &c->obj, obj) < 0) { goto _failed; }
		if(++c->curr == c->count) {
			obj = c->obj;
			--top;
			/*printf("stack pop %d\n", top);*/
			goto _push;
		}
		goto _header_again;
	case CT_MAP_KEY:
		c->map_key = obj;
		c->ct = CT_MAP_VALUE;
		goto _header_again;
	case CT_MAP_VALUE:
		if(msgpack_unpack_callback(_map_item)(user, &c->obj, c->map_key, obj) < 0) { goto _failed; }
		if(--c->count == 0) {
			obj = c->obj;
			--top;
			/*printf("stack pop %d\n", top);*/
			goto _push;
		}
		c->ct = CT_MAP_KEY;
		goto _header_again;

	default:
		goto _failed;
	}

_header_again:
		cs = CS_HEADER;
		++p;
	} while(p != pe);
	goto _out;


_finish:
	stack[0].obj = obj;
	++p;
	ret = 1;
	/*printf("-- finish --\n"); */
	goto _end;

_failed:
	/*printf("** FAILED **\n"); */
	ret = -1;
	goto _end;

_out:
	ret = 0;
	goto _end;

_end:
	ctx->cs = cs;
	ctx->trail = trail;
	ctx->top = top;
	*off = p - (const unsigned char*)data;

	return ret;
}


#undef msgpack_unpack_func
#undef msgpack_unpack_callback
#undef msgpack_unpack_struct
#undef msgpack_unpack_object
#undef msgpack_unpack_user

#undef push_simple_value
#undef push_fixed_value
#undef push_variable_value
#undef again_fixed_trail
#undef again_fixed_trail_if_zero
#undef start_container

#undef NEXT_CS
#undef PTR_CAST_8
#undef PTR_CAST_16
#undef PTR_CAST_32
#undef PTR_CAST_64

